### PR TITLE
chore(ci): Refactor CI to use prod images in staging and prevent accidental hotfix breakages

### PR DESCRIPTION
**Pull request by jamilbk**
- Runs release asset builds simultaneously with `deploy-staging`. Those don't depend on each other.
- Prevents running some build workflows in CD because they're run already in the PR and in the merge group, and the risk of semantic conflict is negligible
- Run `release` assets in staging
- Adds `compatibility_tests`: **To successfully introduce a breaking change in the control / data plane APIs, you must now "Merge as Administrator"**
- Since `CI` is no longer run on `main`, caching needed to be refactored to make sense again
- Since `CI` is no longer run on `main`, the Elixir `migrations_and_seeds_test` had to be rewritten. This now tests migrations using `git checkout` instead of importing `main`'s DB dump.
- Move tauri builds to its own workflow so we can trigger Linux and Windows builds manually on an adhoc basis like we do for the Swift and Kotlin builds
- Add a new `hotfix` workflow that will run `compatibility_tests` with the latest published images
- Add `workflow_dispatch` to trigger `CD` manually for testing purposes (cc @ReactorScram)

Refs #3995

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -11,23 +11,41 @@ concurrency:
...
+
+  # We build tauri artifacts in CD, no need to build it before here.
+  tauri:
+    if: github.workflow == "Continuous Integration"
+    uses: ./.github/workflows/_tauri.yml
+    secrets: inherit
+
+  # These workflows have tests which may catch semantic
+  # merge errors resulting from the final merge to main
+  # that the PR and merge group workflows don't catch.
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]: How? The merge group should prevent semantic conflicts.

File: .github/workflows/ci.yml
Comment (thomaseizinger) [inline .github/workflows/ci.yml:1]: Maybe we should just not run CI on `main` at all?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jamilbk) [reply .github/workflows/ci.yml]: Yeah good point, I thought so, but it was late when I was thinking through this and wasn't 100% sure.

I'll update all of these.

File: .github/workflows/ci.yml
Comment (jamilbk) [reply .github/workflows/ci.yml:1]: Yeah, I'll revise this PR. We may need some of those jobs to run on `main`, but I'll see.

Comment (jamilbk) [reply .github/workflows/ci.yml:1]: Ah, I remember why.

We build all of the Rust components in CI for `debug`. Then when we prepare the release in `cd.yml`, we rebuild all of them for `release` -- and notably, we *do not* re-run the integration tests on these. The production release binaries get shipped with no tests.

I think what we should do is:

- Create a single `_build_artifacts.yml` workflow. This builds both binaries and Docker images. It takes an input, `target=debug` or `target=release`.
- Move `integration-tests` and `snownet-integration-tests` to `_integration_test.yml` workflow. This optionally takes a commit SHA or version tag for control plane and data plane images to run the tests with.
- In `ci.yml`, call the various language workflows, and the `_build_artficats.yml` workflow with `target=debug`, then run `_integration_tests.yml`. Lastly, run `perf-tests` and `compare-results`.
- In `cd.yml`, call just the `_build_artifacts.yml` workflow with `target=release`, then run
  the same `_integration_tests.yml`.
- If everything passes, `deploy-staging`.

When publishing a release, we then just need to:

- Copy the staging docker images in GCR to prod
- `deploy-prod` with these
- Copy the GCR `client` and `gateway` images to `ghcr.io` for public consumption

When performing a hotfix, we can then:

- Run `_integration_test.yml` with the current SHA for control plane images, and the latest SHA for the data plane images on `ghcr.io`
- Run `deploy-production`

**Comment by ReactorScram** [comment]
After this gets in, I would be tempted to do manual mutation tests to validate it, e.g.:
- Open a draft PR with an obvious performance regression, see if the perf tests fail
- PR with compatibility break, see if integration can catch it

It's also been difficult to make any changes to CD since running CD updates the draft release and does a bunch of stuff. So if this means that we can do almost all of CD, such as building the release MSI and AppImage, without the destructive steps from merging into main and deploying to prod, that would be cool.

File: .github/workflows/ci.yml
Comment (thomaseizinger) [reply .github/workflows/ci.yml:1]: Yeah this sounds really great!

**Comment by thomaseizinger** [comment]
> * Open a draft PR with an obvious performance regression, see if the perf tests fail

We don't automatically verify the performance results (yet).

**Comment by jamilbk** [comment]
> It's also been difficult to make any changes to CD since running CD updates the draft release and does a bunch of stuff. So if this means that we can do almost all of CD, such as building the release MSI and AppImage, without the destructive steps from merging into main and deploying to prod, that would be cool.

This should be simple to do with this refactoring, however some things like cache writing only happen on `main` and can't be tested unless run there. I will see if I can update those jobs to write cache on `CD` and not `main` so we can trigger them with a manual run.

Otherwise, CI and CD are essentially the same now except for building `--debug` and building `--release`.

**Comment by jamilbk** [comment]
Performing a bit more fixup / cleanup, will mark for review again in a bit

**Comment by jamilbk** [comment]
@firezone/engineering Ready for review. This is likely to break `main` because the `CD` workflow dispatch trigger isn't available until merged to `main`, so it's a bit hard to test. After this PR "settles" though, we'll be able to do that in the future.

File: .github/actions/setup-elixir/action.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      if: ${{ steps.cache.outputs.cache-hit != 'true' }}
+      with:
+        path: |
+          elixir/deps
+          elixir/_build/${{ inputs.mix_env }}
+        key: ${{ steps.cache.outputs.cache-primary-key }}
+    - name: Compile Application
+      shell: bash
+      working-directory: ./elixir
+      run: mix compile --warnings-as-errors ${{ inputs.force_compile && '--force' || '' }}
```
Comment (thomaseizinger) [inline .github/actions/setup-elixir/action.yml]: More of a semantic thing but compiling the app seems a bit much for a "setup" action.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -31,214 +34,42 @@ jobs:
...
-        run: docker compose logs relay
-      - name: Show Gateway logs
-        if: "!cancelled()"
-        run: docker compose logs gateway
-      - name: Show API logs
-        if: "!cancelled()"
-        run: docker compose logs api
+  compatibility-tests:
+    # Allow breaking changes only if the PR contains a '!' in the title
+    if: github.event_name == 'pull_request' && ! contains(github.event.pull_request.title, '!')
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]: If this doesn't always run, we can't make it a required job, right?

I think I'd prefer if we have to use the "override with admin privileges" checkbox to merge a PR that has breaking changes.

File: .github/workflows/hotfix.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      elixir_tag: ${{ inputs.tag }}
+      # Should match relay version
+      snownet_tag: ${{ inputs.tag }}
+
+      # Use the earliest supported version of the gateway and client.
+      # Today that's the same as the latest published one.
+      gateway_image: ghcr.io/firezone/gateway
+      gateway_tag: latest
+      client_image: ghcr.io/firezone/client
+      client_tag: latest
```
Comment (thomaseizinger) [inline .github/workflows/hotfix.yml]: In the other workflow you mentioned that it is not published yet. Which one is true? :)

Also, if this is the linux CLI client, should we include that in the image name?

File: .github/actions/setup-elixir/action.yml
Comment (thomaseizinger) [inline .github/actions/setup-elixir/action.yml:1]: Could this be landed as a separate PR?

**Review by thomaseizinger — COMMENTED** [review]
I really like the design of this, great work!

Added a few comments. I think where possible, it would be nice to extract some smaller PRs just so we move in smaller steps.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (jamilbk) [reply .github/workflows/ci.yml]: 👍

File: .github/workflows/_build_artifacts.yml [hunk 5]
```diff
@@ -0,0 +1,326 @@
+name: Build Artifacts
```
Comment (ReactorScram) [inline .github/workflows/_build_artifacts.yml:1]: Does the leading underscore in the file name mean that this is only called by other workflows, and isn't top-level?

File: .github/actions/setup-rust/action.yml
Comment (jamilbk) [inline .github/actions/setup-rust/action.yml:1]: I'll revert these changes. I think the GH actions rust cache has a much better chance of caching this (and probably the Kotlin/Swift builds too)

File: .github/actions/setup-elixir/action.yml (hunk 2 above)
Comment (jamilbk) [reply .github/actions/setup-elixir/action.yml]: Fixed in #4067

File: .github/workflows/hotfix.yml (hunk 4 above)
Comment (jamilbk) [reply .github/workflows/hotfix.yml]: Yeah I gave some thought to renaming the image, but settled on just `client` since all Docker images are Linux, and `ghcr.io/firezone/client` is a bit more concise

File: .github/actions/setup-elixir/action.yml
Comment (jamilbk) [reply .github/actions/setup-elixir/action.yml:1]: #4067

File: .github/workflows/_build_artifacts.yml (hunk 5 above)
Comment (jamilbk) [reply .github/workflows/_build_artifacts.yml:1]: It's a convention I think @AndrewDryga started, so I kept it for this refactor too.

File: .github/actions/setup-rust/action.yml
Comment (jamilbk) [reply .github/actions/setup-rust/action.yml:1]: Reverted

File: .github/workflows/hotfix.yml (hunk 4 above)
Comment (jamilbk) [reply .github/workflows/hotfix.yml]: Fixed the hotfix client image! I suspect we'll probably publish another release before this hotfix is run again.

**Comment by jamilbk** [comment]
@firezone/engineering -- I'll need to update the required checks before this gets merged in. After merge, there's guaranteed to be some issues I'll need to work through. I'll prioritize that today to minimize disruption.

This will block merges until you rebase.

File: .github/workflows/_build_artifacts.yml (hunk 5 above)
Comment (AndrewDryga) [reply .github/workflows/_build_artifacts.yml:1]: @ReactorScram I just tried to separate the workflow that we do run from helper files like that we just include because we have a lot of them and GitHub actions can't use subfolders.

File: .github/workflows/_deploy_production.yml [hunk 6]
```diff
@@ -0,0 +1,100 @@
+name: Deploy Production
```
Comment (AndrewDryga) [inline .github/workflows/_deploy_production.yml:1]: I guess would be good to have `_deploy_staing.yml` too.

File: .github/workflows/_deploy_production.yml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    permissions:
+      contents: write
+    env:
+      TF_CLOUD_ORGANIZATION: "firezone"
+      TF_API_TOKEN: "${{ secrets.TF_API_TOKEN }}"
+      TF_WORKSPACE: "production"
+    steps:
+      - name: Get Terraform Version
+        run: |
+          TERRAFORM_VERSION=$(cat .tool-versions | grep terraform | awk '{ print $NF; }')
```
Comment (AndrewDryga) [inline .github/workflows/_deploy_production.yml:81]: We should use:
```
- name: Tool Versions
        id: versions
        uses: marocchino/tool-versions-action@v1.2.0
```

File: .github/workflows/_build_artifacts.yml (hunk 5 above)
Comment (ReactorScram) [reply .github/workflows/_build_artifacts.yml:1]: Makes sense, I just wanted to be sure my guess was right

File: .github/workflows/_deploy_production.yml (hunk 7 above)
Comment (jamilbk) [reply .github/workflows/_deploy_production.yml:81]: Makes sense, I'll open that as a separate PR ontop of this one to keep this one from getting larger

File: .github/workflows/_integration_tests.yml [hunk 8]
```diff
@@ -0,0 +1,161 @@
+name: Integration Tests
```
Comment (AndrewDryga) [inline .github/workflows/_integration_tests.yml:1]: 👍

File: .github/workflows/cd.yml [hunk 9]
```diff
@@ -9,21 +17,27 @@ env:
   VERSION: "1.0.0"
 
 jobs:
-  ci:
```
Comment (AndrewDryga) [inline .github/workflows/cd.yml:12]: Not running CI feels *very* unsafe to me, the whole goal of CI is to make sure that if the `main` is green - everything is fine, it doesn't matter what and when is tested in other branches, PRs, etc. Now we give up this safety for optimization in a repo where anyone can merge as administrator, with a growing team that will fail us sooner than later.

If we want to shovel off time from the release cycle I would do it in PR's or merge queues, but not from `main`..

File: .github/workflows/_deploy_production.yml [hunk 10]
```diff
@@ -0,0 +1,100 @@
+name: Deploy Production
+run-name: Triggered by ${{ github.actor }}
+on:
+  workflow_call:
+    inputs:
+      tag:
+        description: "Image tag to deploy. Defaults to the last commit SHA in the branch."
```
Comment (AndrewDryga) [inline .github/workflows/_deploy_production.yml:7]: I think we can use `gh` to fetch the latest Actions run for the given commit SHA and make sure the CI did pass.

File: .github/workflows/_deploy_production.yml (hunk 6 above)
Comment (jamilbk) [reply .github/workflows/_deploy_production.yml:1]: Hm yeah we could, but I may do that in a separate PR. We only call it from one place, `cd.yml` so that's why I inlined it there.

File: .github/workflows/cd.yml (hunk 9 above)
Comment (jamilbk) [reply .github/workflows/cd.yml:12]: Updated to run CI in CD

File: .github/workflows/_deploy_production.yml (hunk 10 above)
Comment (jamilbk) [reply .github/workflows/_deploy_production.yml:7]: Done

File: .github/workflows/hotfix.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
+name: Hotfix Production
+run-name: Triggered by ${{ github.actor }}
+on:
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: "Image tag to deploy. Defaults to the last commit SHA in the branch."
```
Comment (ReactorScram) [inline .github/workflows/hotfix.yml:7]: 👍  So this can roll back to a SHA older than the tip of main, e.g. if a deployment goes wrong and we want to just quickly undo it? Looks good.

File: docker-compose.yml [hunk 12]
```diff
@@ -56,7 +56,7 @@ services:
         - type=registry,ref=us-east1-docker.pkg.dev/firezone-staging/cache/web:main
       args:
         APPLICATION_NAME: web
-    image: us-east1-docker.pkg.dev/firezone-staging/firezone/web:${VERSION:-main}
+    image: ${WEB_IMAGE:-us-east1-docker.pkg.dev/firezone-staging/firezone/web}:${WEB_TAG:-main}
```
Comment (ReactorScram) [inline docker-compose.yml:59]: 👍  The web, client, and gateway images were split up so that we can do the compat tests, previously everything was just VERSION, right? Looks good.

File: rust/Dockerfile [hunk 13]
```diff
@@ -134,7 +134,8 @@ RUN set -xe \
 FROM development AS debug
 
 ARG TARGET
-COPY --from=builder /build/target/${TARGET}/debug/${PACKAGE} .
+## Build first with `cross build --target ${TARGET} -p ${PACKAGE} --debug && mv /target/${TARGET}/release/${PACKAGE} .`
```
Comment (ReactorScram) [inline rust/Dockerfile:137]: I don't understand the comment here. Is this saying what the Dockerfile did, or what the reader should do to replicate it manually without Docker?

**Review by ReactorScram — APPROVED** [review]
I think I understand it. It looks like everything I asked for :) 👍 

Most of my comments will be questions to make sure I understand everything, since I haven't looked deep into this before and haven't run deployments.

So if I understand it right:
- There are 4 top-level workflows (forgetting e2e), CD, CI, Hotfix, and Publish
- CI and CD are non-destructive by default. Hotfix and Publish are the only ones that change production. (CD can change staging, but if I just need to debug an MSI bundling issue, it doesn't change staging by default.)
- If I want a "same-as-production" MSI or AppImage for the GUI Client, I can trigger `_tauri.yml` manually and the runners won't waste time doing anything else.
- Hotfix can be used for fast rollbacks because it doesn't re-run CI/CD, it checks that CI has succeeded before on the SHA and then deploys it to production as fast as possible
- Publish is the formal release process that does everything, asks CI to build release assets, puts those in a new public release, updates staging, updates prod, etc.
- CI took over the role of building release artifacts from CD, so the duplication between, e.g. Tauri MSI builds, is gone, everything for the Clients builds in CI now. CI still does all the tests. CD and Publish run CI. Hotfix requires CI to _have_ succeeded in the past but never runs CI on its own.
- CI is still responsible for all unit tests, smoke tests, integration tests, and now it also performs compatibility tests with the latest tag published to the container registry.
- When the tip of main updates, CD automatically deploys to staging but not to prod.

I forgot e2e since it didn't change. Do the testbed machines run local Github runners and that's how we test things on them? I have not tried it yet.

File: rust/Dockerfile (hunk 13 above)
Comment (jamilbk) [reply rust/Dockerfile:137]: It's what the reader should do to get the image to package when building locally to test / debug release builds.

**Comment by jamilbk** [comment]
@ReactorScram Close!

> CI took over the role of building release artifacts from CD, so the duplication between, e.g. Tauri MSI builds, is gone, everything for the Clients builds in CI now. CI still does all the tests. CD and Publish run CI. Hotfix requires CI to have succeeded in the past but never runs CI on its own.

`Publish` just checks for CI green now too, and doesn't re-run those tests (it uses the same logic for Hotfix, but publishes data plane images to `ghcr.io` too.)

> Publish is the formal release process that does everything, asks CI to build release assets, puts those in a new public release, updates staging, updates prod, etc.

Publish doesn't do all that -- by the time it runs, we should have already verified everything in the CI workflow for the same SHA that's being published. It simply makes public our new `gateway` and `client` Docker images, and deploys production. This is triggered when we publish the drafted release that's been accumulating the artifacts from `CD`.

File: rust/Dockerfile (hunk 13 above)
Comment (ReactorScram) [reply rust/Dockerfile:137]: Oh right, for running docker-compose with local builds.

File: .github/workflows/cd.yml (hunk 9 above)
Comment (jamilbk) [reply .github/workflows/cd.yml:12]: So I now realize the issue with this -- we will never be able to publish breaking changes since `compatibility_tests` will always be red on main.

I think the workaround could be to disable compatibility testing on `main` and only run on PRs and hotfixes.
